### DISCUSSION TITLE
What about CustomData?

### DISCUSSION THREAD

**Discussion by guguxh** [wxyz-abcd/node-haxball#32] [Q&A]
I'm creating a script using this API, but I don't understand what customData does exactly, can you explain me?

**Comment by wxyz-abcd** [comment]
"customData" is your custom data that you might want to pass around between your plugins. It has to be initialized inside onBefore... room callback. The data is passed into each plugin in the same order that you pass them in the array while creating the room. It is passed into the plugins so that the plugins can read and if necessary modify the customData. At last, onAfter... room callback is called with the latest version of customData. This is the mechanism for all callbacks. This improves the code in memory and performance aspects, for those cases that you would need to do the same thing for each plugin.

So, let's say that I want to write a plugin that will control other players with !control [playerid] command. Then, I want to write another plugin that will move a player to a team with !move [playerid] [teamid]. These two seperate plugins have to share data so that you do not have to split the chat message twice.

For example, there is a built-in customData by default in all host-side received operations. All event messages are by default received by an inner room callback "onBeforeOperationReceived". If it is a chat message and a command, the original chat message is split by white-spaces and returned as customData with an indicator of whether it is a command or not. The example code is as follows:
```js
    this.onBeforeOperationReceived = function(obj, msg){
      if (obj.type != OperationType.SendChat)
        return;
      var m = obj.getValue(msg, "text");
      if (m.startsWith("!")){
        return {
          isCommand: true, 
          data: m.trimEnd().split(" ")
        };
      };
      return {
        isCommand: false
      };
    };
```
